src/util/sighash.rs

```diff
@@ -0,0 +1,836 @@
+        //     sha_amounts (32): the SHA256 of the serialization of all spent output amounts.
+        //     sha_scriptpubkeys (32): the SHA256 of the serialization of all spent output scriptPubKeys.
+        //     sha_sequences (32): the SHA256 of the serialization of all input nSequence.
+        if !anyone_can_pay {
+            self.common_cache().prevouts.consensus_encode(&mut writer)?;
+            self.taproot_cache(prevouts.get_all()?)
+                .amounts
+                .consensus_encode(&mut writer)?;
+            self.taproot_cache(prevouts.get_all()?)
+                .script_pubkeys
+                .consensus_encode(&mut writer)?;
+            self.common_cache()
+                .sequences
+                .consensus_encode(&mut writer)?;
+        }
+
+        // If hash_type & 3 does not equal SIGHASH_NONE or SIGHASH_SINGLE:
+        //     sha_outputs (32): the SHA256 of the serialization of all outputs in CTxOut format.
+        if sighash != SigHashType::None && sighash != SigHashType::Single {
+            self.common_cache().outputs.consensus_encode(&mut writer)?;
+        }
+
+        // * Data about this input:
+        // spend_type (1): equal to (ext_flag * 2) + annex_present, where annex_present is 0
+        // if no annex is present, or 1 otherwise
+        let mut spend_type = 0u8;
+        if annex.is_some() {
+            spend_type |= 1u8;
+        }
+        if script_path.is_some() {
+            spend_type |= 2u8;
+        }
+        spend_type.consensus_encode(&mut writer)?;
+
+        // If hash_type & 0x80 equals SIGHASH_ANYONECANPAY:
+        //      outpoint (36): the COutPoint of this input (32-byte hash + 4-byte little-endian).
+        //      amount (8): value of the previous output spent by this input.
+        //      scriptPubKey (35): scriptPubKey of the previous output spent by this input, serialized as script inside CTxOut. Its size is always 35 bytes.
+        //      nSequence (4): nSequence of this input.
+        if anyone_can_pay {
+            let txin = &self.tx.input[input_index];
+            let previous_output = prevouts.get(input_index)?;
+            txin.previous_output.consensus_encode(&mut writer)?;
+            previous_output.value.consensus_encode(&mut writer)?;
+            previous_output
+                .script_pubkey
+                .consensus_encode(&mut writer)?;
+            txin.sequence.consensus_encode(&mut writer)?;
+        } else {
+            (input_index as u32).consensus_encode(&mut writer)?;
+        }
+
+        // If an annex is present (the lowest bit of spend_type is set):
+        //      sha_annex (32): the SHA256 of (compact_size(size of annex) || annex), where annex
+        //      includes the mandatory 0x50 prefix.
+        if let Some(annex) = annex {
+            let mut enc = sha256::Hash::engine();
+            annex.as_bytes().to_vec().consensus_encode(&mut enc)?;
```

Comment:
This allocation is unnecessary, I'd propose implementing `Encodable` for `Annex` instead.

Reply:
`impl Encodable for Annex ` added in 16a236b